loader/0032-loader-plugins.md

```diff
@@ -0,0 +1,328 @@
+*/
+void addTentativeCandidate(String group, String modId, Version version, CandidateResolver resolver);
+
+/**
+* Adds a rule to the current solver.
+*/
+void addRule(Rule rule);
+
+/**
+* Adds a LoadOption to the current solver. All existing rules will have Rule#onLoadOptionAdded called, and all plugins 
+* will have ILoaderPlugin#onLoadOptionAdded called.
+*
+* If this is an ITentativeLoadOption then it will be removed at the end of the cycle, and handled by whatever plugin 
+* added it.
+*/
+void addOption(LoadOption option);
+
+/**
+* Gets the metadata for a given mod.
+*
+* Because the mods in QuiltLoader only reference fully loaded mods, this method can be used during the mod loading process
+* to get the metadata for any candidates (tentative or not) that were present prior to this cycle.
+*/
+ModMetadata getMetadata(String modId);
+}
+```
+</details>
+
+<details>
+    <summary>CandidateResolver</summary>
+    
+```java
+interface CandidateResolver {
+/**
+* Attempts to resolve a tentative candidate.
+*
+* @return an error message if resolution fails, null otherwise
+*/ 
+@Nullable String resolve();
+}
+```
+</details>
+
+<details>
+    <summary>QuiltLoaderPlugin</summary>
+
+```java
+/**
+* @param <T> the types of resolver this plugin can resolve
+*/
+interface QuiltLoaderPlugin<T extends CandidateResolver>  {
+/**
+* Called once per cycle as the first action in the cycle.
+* 
+* This is where mods can be added with {@link QuiltPluginContext#addCandidate} and
+* {@link QuiltPluginContext#addTentativeCandidate}.
+*/
+default void run(QuiltPluginContext context) {}
+
+/**
+* Called once per cycle after the sat4j solving has finished, but before any resolvers are run.
+*
+* Should NOT invoke the resolvers.
+*
+* @return true if all of the resolvers can be called, false otherwise
+*/
+default boolean canResolve(List<T> resolvers) {
+    return false;
+}
+
+/**
+* Called if loader can't simplify this error down into any of the other error handling methods.
+* @return True if this plugin did something which will solve / change the error in future,
+*         and so loader won't ask any other plugins to solve this.
+*         Loader will temporarily remove this rule so it won't be sent to #handleOtherErrors again in this cycle.
+*         If this returns false then no rules will be removed, and instead loader will assume that
+*         the error has been handled in some other way. (and it will promptly crash if you haven't)
+*/
+default @Nullable Rule handleOtherErrors(QuiltPluginContext context, List<Rule> errorChain) { return null; }
+
+/**
+* @param dep The dependency which is missing completely. If you can find a valid source for this then you should add 
+*            it with {@link QuiltContext#addTentativeCandidate()}
+* @return True if this plugin did something which will solve / change the error in future,
+*         and so loader won't ask any other plugins to solve this.
```

Comment:
With no consistent plugin load order does this not open up the possibility for inconsistent loading behaviour depending on which plugin was fixing errors? Above you mentioned all plugins getting the same information each cycle; would it not be better to find out if there are multiple plugins which all claim to be able to fix an error, in which case you'd either have to pick one consistently or crash? Similar in a way to how cpw's ModLauncher voting doesn't allow all transformers to pass a cycle, at least one must be happy to run every time.

Reply:
> would it not be better to find out if there are multiple plugins which all claim to be able to fix an error, in which case you'd either have to **pick one consistently** or crash

This would probably be a better solution, but picking one consistently is the difficult part here. I suppose with loader plugins I wouldn't have much of an issue with adding an integer priority and determining based off that, but a more elegant (and less potentially abusable) system would be preferable.